definitions/procedures/backup/config_files.rb

```diff
@@ -8,26 +11,39 @@ class ConfigFiles < ForemanMaintain::Procedure
           Checks::Backup::CertsTarExist.new
         end
       end
+
       param :backup_dir, 'Directory where to backup to', :required => true
       param :proxy_features, 'List of proxy features to backup (default: all)',
             :array => true, :default => ['all']
       param :ignore_changed_files, 'Should packing tar ignore changed files',
             :flag => true, :default => false
     end
 
+    # rubocop:disable Metrics/MethodLength
     def run
-      tarball = File.join(@backup_dir, 'config_files.tar.gz')
-      increments = File.join(@backup_dir, '.config.snar')
-      with_spinner('Collecting config files to backup') do
-        configs, to_exclude = config_files
-        feature(:tar).run(
-          :command => 'create', :gzip => true, :archive => tarball,
-          :listed_incremental => increments, :ignore_failed_read => true,
-          :exclude => to_exclude, :allow_changing_files => @ignore_changed_files,
-          :files => configs.join(' ')
-        )
+      logger.debug("Invoking tar from #{FileUtils.pwd}")
+      tar_cmd = tar_command
+      attempt_no = 1
+      loop do
+        runner = nil
+        with_spinner('Collecting config files to backup') do
+          runner = execute_runner(tar_cmd, :valid_exit_statuses => [0, 1])
+        end
+        break if runner.exit_status == 0 || @ignore_changed_files
+
+        puts "WARNING: Attempt #{attempt_no}/#{MAX_RETRIES} to collect all config files failed!"
+        puts 'Some files were modified during creation of the archive.'
+        if attempt_no == MAX_RETRIES
+          raise runner.execution_error
+        else
+          attempt_no += 1
+          FileUtils.rm_rf(tarball_path)
+          puts "We will re-try after #{RETRY_DELAY} seconds."
```

Comment:
@kgaikwad I think with_content? is useful information about the proxy and the proxy itself should be responsible for providing it. If we start to define content differently e.g. Pulp 3 has different feature or is not mandatory we want to have single place to update.

Reply:
Currently, inside `with_content?` it just verifies the feature presence and method defined inside feature `instance` at [line 43](https://github.com/theforeman/foreman_maintain/blob/bcb52d0032a17b2e8d3c5fd7257695c6fcecb273/definitions/features/instance.rb#L43) says `foreman_proxy_with_content?` so I thought it's better to add condition there.

Secondly, with the different features for Pulp2 & Pulp3 and the suggested way on [PR-290](https://github.com/theforeman/foreman_maintain/pull/290#issuecomment-554988291) it will resolve the concern. What do you think?

Reply:
IMO `feature(:foreman_proxy).with_content?` and `feature(:instance).foreman_proxy_with_content?` are two different things. The former tells if there is content managed with the proxy while the later tells if the machine is proxy with content only (no server). So I think both make sense to keep. 

Further I'd suggest not to reduce with_content to with_pulp without revisiting the referencing code and checking what actually we are testing. I can imagine content proxy that is using remote pulp - it would still be content proxy but the current test will not be sufficient.  If we need just pulp we should test for presence of pulp and not content proxy. Revisiting this is probably out of scope of this PR.

Reply:
Thank you @mbacovsky for giving pointers.
Because after seeing at the method, I was under an impression that it just checks the presence of pulp and didn't get the scope of it.